scripts/ingests/utils.py

```diff
@@ -1,10 +1,253 @@
+        return
+
+    not_null_pub_filters = []
+    if name:
+        fuzzy_query_name = '%' + name + '%'
+        not_null_pub_filters.append(db.Publications.c.name.ilike(fuzzy_query_name))
+    if doi:
+        not_null_pub_filters.append(db.Publications.c.doi.ilike(doi))
+    if bibcode:
+        not_null_pub_filters.append(db.Publications.c.bibcode.ilike(bibcode))
+
+    if len(not_null_pub_filters) > 0:
+        pub_search_table = db.query(db.Publications).filter(or_(*not_null_pub_filters)).table()
+
+    n_pubs_found = len(pub_search_table)
+
+    if n_pubs_found == 1:
+        verboseprint(f'Found {n_pubs_found} matching publications for {name} or {doi} or {bibcode}')
+        if verbose:
+            pub_search_table.pprint_all()
+        return True
+
+    if n_pubs_found > 1:
+        print(f'Found {n_pubs_found} matching publications for {name} or {doi} or {bibcode}')
+        if verbose:
+            pub_search_table.pprint_all()
+        return pub_search_table
+
+    # If no matches found, search using first four characters of input name
+    if n_pubs_found == 0 and name:
+        shorter_name = name[:4]
+        print(f'No matching publications for {name}, Trying {shorter_name}')
+        fuzzy_query_shorter_name = '%' + shorter_name + '%'
+        pub_search_table = db.query(db.Publications).filter(db.Publications.c.name.ilike(fuzzy_query_shorter_name)).table()
+        n_pubs_found_short = len(pub_search_table)
+        if n_pubs_found_short == 0:
+            print(f'No matching publications for {shorter_name}')
+            print('Use add_publication() to add it to the database.')
+            return False
+
+        if n_pubs_found_short > 0:
+            print(f'Found {n_pubs_found_short} matching publications for {shorter_name}')
+            if verbose:
+                pub_search_table.pprint_all()
+            return pub_search_table
+
+    return
+
+
+
+def add_publication(db, doi: str = None, bibcode: str = None, name: str = None, description: str = None, dryrun: bool = True):
+    """
+    Adds publication to the database using DOI or ADS Bibcode, including metadata found with ADS.
+
+    In order to auto-populate the fields, An $ADS_TOKEN environment variable must be set.
+    See https://ui.adsabs.harvard.edu/user/settings/token
+
+    Parameters
+    ----------
+    db
+        Database object
+    doi, bibcode: str
+        The DOI or ADS Bibcode of the reference. One of these is required input.
+    name: str, optional
+        The publication shortname, otherwise it will be generated [optional]
+    description: str, optional
+        Description of the paper, typically the title of the papre [optional]
+    dryrun: bool
+
+    See Also
+    --------
+    search_publication: Function to find publications in the database
+
+    """
+
+    if not(doi or bibcode):
+       print('DOI or Bibcode is required input')
+       return
+
+    ads.config.token = os.getenv('ADS_TOKEN')
```

Comment:
Should we error out if ADS_TOKEN is not set?

Reply:
No, I want the function to still work even without ADS. I'll add more feedback though.